File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;  // Get the length of the input array
        int[] ans = new int[n];  // Initialize the result array to store the final answer
        
        // First pass: Calculate the left product for each element
        // 'left' holds the product of all elements to the left of the current element
        for (int i = 0, left = 1; i < n; ++i) {
            ans[i] = left;  // Store the current left product in the result array
            left *= nums[i];  // Update the left product by multiplying with the current element
        }
        
        // Second pass: Calculate the right product for each element
        // 'right' holds the product of all elements to the right of the current element
        for (int i = n - 1, right = 1; i >= 0; --i) {
            ans[i] *= right;  // Multiply the stored left product with the right product
            right *= nums[i];  // Update the right product by multiplying with the current element
        }
        
        // Return the result array which contains the product of all elements except self
        return ans;
    }
}

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)  # Get the length of the input list
        ans = [0] * n  # Initialize the result list with zeroes
        
        # Initialize left and right products
        left = right = 1  # Both left and right products start from 1
        
        # First pass: calculate the left product for each element
        for i, x in enumerate(nums):  # Iterate over the list
            ans[i] = left  # Store the current left product in the result list
            left *= x  # Update the left product by multiplying with the current element
        
        # Second pass: calculate the right product for each element
        for i in range(n - 1, -1, -1):  # Iterate backward over the list
            ans[i] *= right  # Multiply the current result by the right product
            right *= nums[i]  # Update the right product by multiplying with the current element
        
        # Return the result list which contains the product of all elements except self
        return ans
